### PR TITLE
copr: Vecotorize `concat_ws`

### PR DESCRIPTION
**Pull request by cireu**
PCP #5751 

###  What have you changed?

Vectorize `concat_ws`
###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?
Please select the tests that you ran to verify your changes:
- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

File: components/tidb_query/src/rpn_expr/impl_string.rs [hunk 1]
```diff
@@ -39,6 +39,25 @@ pub fn concat(args: &[&Option<Bytes>]) -> Result<Option<Bytes>> {
     Ok(Some(output))
 }
 
+#[rpn_fn(varg, min_args = 2)]
+#[inline]
+pub fn concat_ws(args: &[&Option<Bytes>]) -> Result<Option<Bytes>> {
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_string.rs]: According to TiDB's implementation I think this function accepts multiple parameters?

Comment (cireu) [reply components/tidb_query/src/rpn_expr/impl_string.rs]: I try `select concat_ws('a)` on MySQL and it throw an incorrect params count error to me.

File: components/tidb_query/src/rpn_expr/impl_string.rs [hunk 2]
```diff
@@ -202,6 +221,65 @@ mod tests {
...
+                vec![
+                    Some(b",".to_vec()),
+                    Some(b"abc".to_vec()),
+                    Some(b"defg".to_vec()),
+                ],
+                Some(b"abc,defg".to_vec()),
+            ),
+            (
+                vec![
+                    Some(b",".to_vec()),
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_string.rs]: Could you add some tests for separators that is not ","?

File: components/tidb_query/src/rpn_expr/impl_string.rs [hunk 3]
```diff
@@ -39,6 +39,25 @@ pub fn concat(args: &[&Option<Bytes>]) -> Result<Option<Bytes>> {
...
+#[inline]
+pub fn concat_ws(args: &[&Option<Bytes>]) -> Result<Option<Bytes>> {
+    if let (Some(sep), Some(s1)) = (args[0], args[1]) {
+        let mut output = s1.to_vec();
+        for s in &args[2..] {
+            if let Some(s) = s {
+                output.extend_from_slice(sep);
+                output.extend_from_slice(s);
+            } else {
+                return Ok(None);
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_string.rs]: This behaviour might not be correct.. According to MySQL when there are nulls the result is not null, which is very interesting.

```
mysql> select concat_ws(',', 'a', null, 'c');
--------------
select concat_ws(',', 'a', null, 'c')
--------------

+--------------------------------+
| concat_ws(',', 'a', null, 'c') |
+--------------------------------+
| a,c                            |
+--------------------------------+
1 row in set (0.04 sec)
```

Comment (cireu) [reply components/tidb_query/src/rpn_expr/impl_string.rs]: Copy that :)

Comment (cireu) [reply components/tidb_query/src/rpn_expr/impl_string.rs]: https://github.com/tikv/tikv/pull/5999 Create an PR here to fix this bug.

Comment (mapleFU) [reply components/tidb_query/src/rpn_expr/impl_string.rs]: Can we testing and merging it after #5999 is merged?

**Comment by mapleFU** [comment]
You can request me to review this pr if you finish it. (>^ω^<)

**Comment by cireu** [comment]
@sre-bot /run-integration-copr-test test=pr/65

**Comment by cireu** [comment]
@mapleFU  Please review ;)

**Comment by cireu** [comment]
Don't merge now.

**Comment by mapleFU** [comment]
I'm sorry! Please create a pr to fix the bug of #5999 first. You can request me review your code.

File: components/tidb_query/src/rpn_expr/impl_string.rs [hunk 4]
```diff
@@ -40,6 +40,25 @@ pub fn concat(args: &[&Option<Bytes>]) -> Result<Option<Bytes>> {
...
 }
 
+#[rpn_fn(varg, min_args = 2)]
+#[inline]
+pub fn concat_ws(args: &[&Option<Bytes>]) -> Result<Option<Bytes>> {
+    if let Some(sep) = args[0] {
+        let mut output = Bytes::new();
+        let rest = &args[1..];
+        let iter = rest.iter().filter_map(|x| x.as_ref()).enumerate();
+        for (idx, x) in iter {
```
Comment (mapleFU) [inline components/tidb_query/src/rpn_expr/impl_string.rs]: You can use `join` to simplify the code.

File: components/tidb_query/src/rpn_expr/impl_string.rs [hunk 5]
```diff
@@ -265,6 +284,90 @@ mod tests {
...
+                vec![Some(b",".to_vec()), Some(b"abc".to_vec())],
+                Some(b"abc".to_vec()),
+            ),
+            (
+                vec![Some(b",".to_vec()), None, Some(b"abc".to_vec())],
+                Some(b"abc".to_vec()),
+            ),
+            (
+                vec![
+                    Some(b",".to_vec()),
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_string.rs]: Please also add null cases like `select concat_ws(',', null, null, 'a', null);`

Comment (cireu) [reply components/tidb_query/src/rpn_expr/impl_string.rs]: https://github.com/tikv/tikv/pull/5978/files#diff-986dc60794ac57d647b1629b74fe123fR388

Add one at here.

File: components/tidb_query/src/rpn_expr/impl_string.rs (hunk 4 above)
Comment (cireu) [reply components/tidb_query/src/rpn_expr/impl_string.rs]: done.

**Comment by cireu** [comment]
@sre-bot /run-integration-copr-test test=pr/65

**Pull request merged**